src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryClientRegistryImpl.java

```diff
@@ -41,23 +39,30 @@ protected CloudFoundryClientRegistryImpl() {
     public CloudFoundryOperations getCloudFoundryClient(ConfigBag conf, boolean allowReuse) {
         String username = checkNotNull(conf.get(PaasLocationConfig.ACCESS_IDENTITY), "identity must not be null");
         String password = checkNotNull(conf.get(PaasLocationConfig.ACCESS_CREDENTIAL), "credential must not be null");
-        URL apiHost = getTargetURL(checkNotNull(conf.get(PaasLocationConfig.CLOUD_ENDPOINT), "endpoint must not be null"));
+        String apiHost = checkNotNull(conf.get(PaasLocationConfig.CLOUD_ENDPOINT), "endpoint must not be null");
         String organization = checkNotNull(conf.get(CloudFoundryPaasLocationConfig.CF_ORG), "organization must not be null");
         String space = checkNotNull(conf.get(CloudFoundryPaasLocationConfig.CF_SPACE), "space must not be null");
 
-        CloudCredentials credentials = new CloudCredentials(username, password);
-        CloudFoundryClient client = new CloudFoundryClient(credentials, apiHost, organization, space, true);
-        client.login();
+        DefaultConnectionContext connectionContext = DefaultConnectionContext.builder()
+                .apiHost(apiHost)
+                .build();
+        //TODO: using singleton for the class instance
+        PasswordGrantTokenProvider tokenProvider = PasswordGrantTokenProvider.builder()
```

Comment:
@andreaturli just a couple of things about the supplier

I have added a little supplier to provide the operations.
This solution looks thread-safe. Only a `token` and a `context` are used to build all the `CloudFoundryClient` instances. 

We may use something `memorize` method but the issue is that we need a lazy initialization here, because the required data to build the client are not known by the client until the method `getClient(ConfigBag config, boolean allowedReuse)` is called.
We could add a `init` method tot he registry and ensure that is called from `CloudFoundryPaasLocation#init()` it will be in charge of initialized the supplier. But I am not sure it it is very smart...